In [4]:
include("utils/data_loader.jl")
include("utils/visualization.jl")
include("utils/ml1_utils.jl")
include("utils/preprocessing.jl")

"    # ONE-HOT ENCODING\n    function oneHotEncoding(feature::AbstractArray{<:Any,1},      \n        classes::AbstractArray{<:Any,1})\n    # First we are going to set a line as defensive to check values\n    @assert(all([in(value, classes) for value in feature]));\n    \n    #" ⋯ 9683 bytes ⋯ "ights)\n            WeightedAccuracy = sum(Accuracy .* Weights) / sum(Weights)\n            return (Sensitivity, Specificity, PPV, NPV, F1, Accuracy, WeightedSensitivity, WeightedSpecificity, WeightedPPV, WeightedNPV, WeightedF1, WeightedAccuracy)\n        end\n    end\nend\n"

In [2]:
using Random
Random.seed!(123)

data = DataLoader.load_data_for_analysis("dataset\\star_classification.csv");
#Visualization.entry_visualization(data);

### Preprocessing ###

In [5]:
using DataFrames

data = DataLoader.load_data("dataset\\star_classification.csv");

# Preprocess the data

"""    This function does the following:
        - Balance the data using the undersampling method if chosen to do so
        - Parse the data: chosing the correct columns for inputs and targets (Shouldn't this be done before balancing??)
        - Splits the data into training and testing using holdOut method
        - Normalize the inputs  ------> zeromean method still to be implemented!!!!
"""
# preprocess_data(data, train_ratio, norm_method, balance, indices)
train_inputs, train_targets, test_inputs, test_targets = Preprocessing.preprocess_data(data, 0.1, "minmax", true, [4,5,6,7,8,15]);

""" I've changed the preprocess_data function so it doesn't OneHotEncode the targets
    because it's not needed&advised for KNN, DT & SVM, only for ANN. 
    The OneHotEncoding for the ANN will be done in the modelCrossValidation function,
    which is called by the evaluateAndPrintMetricsRanking function.
"""

Train_inputs before normalization: Float32[18.85256, 18.40885, 18.28408, 18.31415, 18.38598, 0.00333416]


" I've changed the preprocess_data function so it doesn't OneHotEncode the targets\n    because it's not needed&advised for KNN, DT & SVM, only for ANN. \n    The OneHotEncoding for the ANN will be done in the modelCrossValidation function,\n    which is called by the evaluateAndPrintMetricsRanking function.\n"

In [8]:
using StatsBase

counts = countmap(train_targets)
total = length(train_targets)

proportions = Dict(k => v / total for (k, v) in counts)


Dict{SubString{String}, Float64} with 3 entries:
  "GALAXY" => 0.334193
  "QSO"    => 0.332474
  "STAR"   => 0.333333

In [4]:
println(typeof(train_inputs), typeof(train_targets),typeof(test_inputs), typeof(test_targets))
println(size(train_inputs),size(train_targets),size(test_inputs),size(test_targets))
println(train_inputs[1:10])

Matrix{Float32}Vector{Any}Matrix{Float32}Vector{Any}
(569, 6)(569,)(56314, 6)(56314,)
Float32[0.29855812, 0.1377225, 0.3577274, 0.71487767, 0.42786127, 0.5657304, 0.6857209, 0.45806453, 0.7845978, 0.84025544]


In [5]:
# Check size of train and test sets
println("Train inputs: ", size(train_inputs))
println("Train targets: ", size(train_targets))
println("Test inputs: ", size(test_inputs))
println("Test targets: ", size(test_targets))

Train inputs: (569, 6)
Train targets: (569,)
Test inputs: (56314, 6)
Test targets: (56314,)


### Testing hyperparameters for each model ###

In [6]:
using ScikitLearn

@sk_import neural_network: MLPClassifier
@sk_import svm: SVC
@sk_import tree: DecisionTreeClassifier
@sk_import neighbors: KNeighborsClassifier
@sk_import ensemble: StackingClassifier  # For the stacking ensemble
@sk_import linear_model: LogisticRegression  # For the final estimator in stacking

PyObject <class 'sklearn.linear_model._logistic.LogisticRegression'>

In [7]:
"""
Setting indices for the k-fold cross-validation
    we are about to do with the different models
"""
N=size(train_inputs,1)
k = 5 # number of folds
kFoldIndices = crossvalidation(N, k);

#### Decision Tree ####

In [8]:
# Define an array of hyperparameter dictionaries for the Decision Tree model
dtree_hyperparameters_array = [
    Dict("max_depth" => 3),
    Dict("max_depth" => 5),
    Dict("max_depth" => 10),
    Dict("max_depth" => 20),
    Dict("max_depth" => 50),
    Dict("max_depth" => 100) # Deeper trees can capture more detail but risk overfitting
]

# Call the function to evaluate the model using different sets of hyperparameters and print the ranking of metrics.
evaluateAndPrintMetricsRanking(:DecisionTree,dtree_hyperparameters_array, train_inputs, train_targets, kFoldIndices)


Training with set of hyperparameters 1
Training with set of hyperparameters 2
Training with set of hyperparameters 3
Training with set of hyperparameters 4
Training with set of hyperparameters 5
Training with set of hyperparameters 6

----- acc -----
Set of hyperparameters 2 -> mean: 0.947 Std. Dev.: 0.02
Set of hyperparameters 3 -> mean: 0.944 Std. Dev.: 0.016
Set of hyperparameters 4 -> mean: 0.944 Std. Dev.: 0.016
Set of hyperparameters 5 -> mean: 0.944 Std. Dev.: 0.016
Set of hyperparameters 6 -> mean: 0.944 Std. Dev.: 0.016
Set of hyperparameters 1 -> mean: 0.921 Std. Dev.: 0.016

----- sensitivity -----
Set of hyperparameters 2 -> mean: 0.947 Std. Dev.: 0.02
Set of hyperparameters 3 -> mean: 0.944 Std. Dev.: 0.016
Set of hyperparameters 4 -> mean: 0.944 Std. Dev.: 0.016
Set of hyperparameters 5 -> mean: 0.944 Std. Dev.: 0.016
Set of hyperparameters 6 -> mean: 0.944 Std. Dev.: 0.016
Set of hyperparameters 1 -> mean: 0.921 Std. Dev.: 0.016

----- specificity -----
Set of hyperparam

#### KNN ####

In [9]:
# Define an array of hyperparameter dictionaries for the kNN model
knn_hyperparameters_array = [
    Dict("n_neighbors" => 5),
    Dict("n_neighbors" => 10),
    Dict("n_neighbors" => 15),
    Dict("n_neighbors" => 20),
    Dict("n_neighbors" => 50),
    Dict("n_neighbors" => 100) # Large neighborhoods, smooths out predictions
]

# Call the function to evaluate the model using different sets of hyperparameters and print the ranking of metrics.
evaluateAndPrintMetricsRanking(:kNN,knn_hyperparameters_array, train_inputs, train_targets, kFoldIndices)

Training with set of hyperparameters 1
Training with set of hyperparameters 2
Training with set of hyperparameters 3
Training with set of hyperparameters 4
Training with set of hyperparameters 5
Training with set of hyperparameters 6

----- acc -----
Set of hyperparameters 1 -> mean: 0.881 Std. Dev.: 0.024
Set of hyperparameters 3 -> mean: 0.845 Std. Dev.: 0.03
Set of hyperparameters 2 -> mean: 0.842 Std. Dev.: 0.036
Set of hyperparameters 4 -> mean: 0.826 Std. Dev.: 0.042
Set of hyperparameters 5 -> mean: 0.787 Std. Dev.: 0.041
Set of hyperparameters 6 -> mean: 0.75 Std. Dev.: 0.03

----- sensitivity -----
Set of hyperparameters 1 -> mean: 0.881 Std. Dev.: 0.024
Set of hyperparameters 3 -> mean: 0.845 Std. Dev.: 0.03
Set of hyperparameters 2 -> mean: 0.842 Std. Dev.: 0.036
Set of hyperparameters 4 -> mean: 0.826 Std. Dev.: 0.042
Set of hyperparameters 5 -> mean: 0.787 Std. Dev.: 0.041
Set of hyperparameters 6 -> mean: 0.75 Std. Dev.: 0.03

----- specificity -----
Set of hyperparameter

#### ANN ####

In [10]:
# Define an array of hyperparameter dictionaries for the ANN model
ann_hyperparameters_array = [
    # Two-layer architecture, moderate neurons
    Dict("architecture" => [50, 30], "activation" => "relu", "learning_rate" => 0.01, "validation_ratio" => 0.1, "n_iter_no_change" => 80, "max_iter" => 1000, "repetitionsTraining" => 10),

    # One-layer architecture, fewer neurons
    Dict("architecture" => [30], "activation" => "relu", "learning_rate" => 0.01, "validation_ratio" => 0.1, "n_iter_no_change" => 80, "max_iter" => 1000, "repetitionsTraining" => 10),

    # Two-layer, different activation function
    Dict("architecture" => [50, 30], "activation" => "tanh", "learning_rate" => 0.01, "validation_ratio" => 0.1, "n_iter_no_change" => 80, "max_iter" => 1000, "repetitionsTraining" => 10),

    # One-layer, lower learning rate
    Dict("architecture" => [30], "activation" => "relu", "learning_rate" => 0.001, "validation_ratio" => 0.1, "n_iter_no_change" => 80, "max_iter" => 2000, "repetitionsTraining" => 10),

    # Two-layer, higher learning rate
    Dict("architecture" => [50, 30], "activation" => "relu", "learning_rate" => 0.05, "validation_ratio" => 0.1, "n_iter_no_change" => 80, "max_iter" => 1000, "repetitionsTraining" => 10),

    # One-layer, logistic activation
    Dict("architecture" => [30], "activation" => "logistic", "learning_rate" => 0.01, "validation_ratio" => 0.1, "n_iter_no_change" => 80, "max_iter" => 1000, "repetitionsTraining" => 10),

    # Two-layer, more neurons, different activation
    Dict("architecture" => [70, 40], "activation" => "tanh", "learning_rate" => 0.01, "validation_ratio" => 0.1, "n_iter_no_change" => 80, "max_iter" => 1000, "repetitionsTraining" => 10),

    # One-layer, more neurons
    Dict("architecture" => [50], "activation" => "relu", "learning_rate" => 0.01, "validation_ratio" => 0.1, "n_iter_no_change" => 80, "max_iter" => 1000, "repetitionsTraining" => 10)
]

# Call the function to evaluate the model using different sets of hyperparameters and print the ranking of metrics.
evaluateAndPrintMetricsRanking(:ANN, ann_hyperparameters_array, train_inputs, train_targets, kFoldIndices)

Training with set of hyperparameters 1
Training with set of hyperparameters 2
Training with set of hyperparameters 3
Training with set of hyperparameters 4
Training with set of hyperparameters 5
Training with set of hyperparameters 6
Training with set of hyperparameters 7
Training with set of hyperparameters 8

----- acc -----
Set of hyperparameters 5 -> mean: 0.93 Std. Dev.: 0.014
Set of hyperparameters 7 -> mean: 0.93 Std. Dev.: 0.014
Set of hyperparameters 1 -> mean: 0.93 Std. Dev.: 0.015
Set of hyperparameters 3 -> mean: 0.929 Std. Dev.: 0.015
Set of hyperparameters 8 -> mean: 0.908 Std. Dev.: 0.024
Set of hyperparameters 2 -> mean: 0.905 Std. Dev.: 0.019
Set of hyperparameters 6 -> mean: 0.829 Std. Dev.: 0.041
Set of hyperparameters 4 -> mean: 0.394 Std. Dev.: 0.047

----- sensitivity -----
Set of hyperparameters 5 -> mean: 0.93 Std. Dev.: 0.014
Set of hyperparameters 7 -> mean: 0.93 Std. Dev.: 0.014
Set of hyperparameters 1 -> mean: 0.93 Std. Dev.: 0.015
Set of hyperparameters 3 

#### SVM ####

In [11]:
svm_hyperparameters_array = [
    # Uses 'rbf' kernel, medium complexity with C=1.0, default polynomial degree, 'scale' for gamma 
    Dict("kernel" => "rbf", "degree" => 3, "C" => 1.0, "gamma" => "scale"),
    
    # Same 'rbf' kernel, increased penalty (C=10.0) for larger-margin separation, 'auto' gamma adjusts based on features
    Dict("kernel" => "rbf", "degree" => 3, "C" => 10.0, "gamma" => "auto"),
    
    # Same 'rbf' kernel, lower penalty (C=0.1) for a softer-margin, 'scale' gamma is default scaling
    Dict("kernel" => "rbf", "degree" => 3, "C" => 0.1, "gamma" => "scale"),

    # 'linear' kernel, suitable for less complex data
    Dict("kernel" => "linear", "degree" => 3, "C" => 0.1, "gamma" => "auto"),
    
    # 'linear' kernel, with C=1.0 indicating a balance between margin and misclassification
    Dict("kernel" => "linear", "degree" => 3, "C" => 1.0, "gamma" => "auto"),

    # 'linear' kernel, with a medium penalty and scale gamma
    Dict("kernel" => "linear", "degree" => 3, "C" => 1.0, "gamma" => "scale"),

    # 'linear' kernel with a higher penalty and scale gamma
    Dict("kernel" => "linear", "degree" => 3, "C" => 10.0, "gamma" => "scale"),
    
    # 'poly' kernel, polynomial degree is set twice by mistake, should only be 'degree' => 3, 'scale' gamma defaults to feature scale
    Dict("kernel" => "poly", "degree" => 3, "C" => 1.0, "gamma" => "scale"),
    
    # 'poly' kernel, increased polynomial degree (5) for higher model complexity, 'auto' gamma may overfit with high dimension
    Dict("kernel" => "poly", "degree" => 5, "C" => 1.0, "gamma" => "auto")
]


# Call the function to evaluate the model using different sets of hyperparameters and print the ranking of metrics.
evaluateAndPrintMetricsRanking(:SVM, svm_hyperparameters_array, train_inputs, train_targets, kFoldIndices)

Training with set of hyperparameters 1
Training with set of hyperparameters 2
Training with set of hyperparameters 3
Training with set of hyperparameters 4
Training with set of hyperparameters 5
Training with set of hyperparameters 6
Training with set of hyperparameters 7
Training with set of hyperparameters 8
Training with set of hyperparameters 9

----- acc -----
Set of hyperparameters 8 -> mean: 0.9 Std. Dev.: 0.029
Set of hyperparameters 1 -> mean: 0.879 Std. Dev.: 0.024
Set of hyperparameters 7 -> mean: 0.87 Std. Dev.: 0.016
Set of hyperparameters 2 -> mean: 0.861 Std. Dev.: 0.027
Set of hyperparameters 5 -> mean: 0.793 Std. Dev.: 0.033
Set of hyperparameters 6 -> mean: 0.793 Std. Dev.: 0.033
Set of hyperparameters 3 -> mean: 0.772 Std. Dev.: 0.022
Set of hyperparameters 4 -> mean: 0.663 Std. Dev.: 0.051
Set of hyperparameters 9 -> mean: 0.353 Std. Dev.: 0.066

----- sensitivity -----
Set of hyperparameters 8 -> mean: 0.9 Std. Dev.: 0.029
Set of hyperparameters 1 -> mean: 0.879 St

In [1]:
"
using Plots

# Define the data for each model
ann_means = [0.947, 0.947, 0.925, 0.788, 0.948, 0.933, 0.8, 0.9]
ann_stds = [0.018, 0.04, 0.07, 0.097, 0.036, 0.039, 0.4, 0.3]
svm_means = [0.947, 0.947, 0.927, 0.953, 0.94, 0.4, 0.5]
svm_stds = [0.03, 0.038, 0.092, 0.051, 0.043, 0.082, 0.7]
dt_means = [0.927, 0.913, 0.913, 0.913, 0.913, 0.913]
dt_stds = [0.043, 0.045, 0.045, 0.045, 0.045, 0.045]
knn_means = [0.947, 0.947, 0.96, 0.94, 0.913, 0.507]
knn_stds = [0.038, 0.038, 0.015, 0.043, 0.104, 0.068]

# Create subplots for each model
p1 = bar(1:6, ann_means, yerr=ann_stds, title="ANN", legend=false)
p2 = bar(1:6, svm_means, yerr=svm_stds, title="SVM", legend=false)
p3 = bar(1:6, dt_means, yerr=dt_stds, title="Decision Tree", legend=false)
p4 = bar(1:6, knn_means, yerr=knn_stds, title="KNN", legend=false)

# Customize the y-axis and labels
for p in [p1, p2, p3, p4]
    ylabel!(p, "Accuracy")
    xlabel!(p, "Set of Hyperparameters")
end

# Combine the plots into one figure
plot(p1, p2, p3, p4, layout=(2,2), size=(800,600))
"

LoadError: bar recipe: x must be same length as y (centers), or one more than y (edges).
		length(x)=6, length(y)=8

In [37]:
using JLD

dt_model = DecisionTreeClassifier(max_depth=10)

# Fit the model on the training data
fit!(dt_model, train_inputs, train_targets)

# Predict the targets for the test data
predicted_targets = predict(dt_model, test_inputs)

# Calculate and print the accuracy
dt_model_accuracy = mean(predicted_targets .== test_targets)
println("Decision Tree model accuracy: $(dt_model_accuracy * 100) %")

# Save the model
#JLD.save("dt_model.jld", "model", dt_model)

Decision Tree model accuracy: 85.74777142451255 %


In [13]:
knn_model = KNeighborsClassifier(n_neighbors=10)

# Fit the model on the training data
fit!(knn_model, train_inputs, train_targets)

# Predict the targets for the test data
predicted_targets = predict(knn_model, test_inputs)

# Calculate and print the accuracy
knn_model_accuracy = mean(predicted_targets .== test_targets)
println("KNN model accuracy: $(knn_model_accuracy * 100) %")

# Save the model
#JLD.save("knn_model.jld", "model", knn_model)

KNN model accuracy: 36.65695919309586 %


In [14]:
ann_model = MLPClassifier(hidden_layer_sizes=(70, 40), activation="tanh", learning_rate_init=0.01, validation_fraction=0.1, n_iter_no_change=80, max_iter=1000)

# Fit the model on the training data
fit!(ann_model, train_inputs, train_targets)

# Predict the targets for the test data
predicted_targets = predict(ann_model, test_inputs)

# Calculate and print the accuracy
ann_model_accuracy = mean(predicted_targets .== test_targets)
println("ANN model accuracy: $(ann_model_accuracy * 100) %")

# Save the model
#JLD.save("ann_model.jld", "model", ann_model)

ANN model accuracy: 74.74340306140569 %


C:\Users\coros\.julia\conda\3\x86_64\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(


In [15]:
svm_model = SVC(kernel="rbf", degree=3, C=1.0, gamma="scale")

# Fit the model on the training data
fit!(svm_model, train_inputs, train_targets)

# Predict the targets for the test data
predicted_targets = predict(svm_model, test_inputs)

# Calculate and print the accuracy
knn_model_accuracy = mean(predicted_targets .== test_targets)
println("SVM model accuracy: $(knn_model_accuracy * 100) %")

# Save the model
#JLD.save("svm_model.jld", "model", svm_model)

SVM model accuracy: 49.40334552686721 %


In [16]:
# Define the base models with the chosen hyperparameters
dt_model = DecisionTreeClassifier(max_depth=10)
knn_model = KNeighborsClassifier(n_neighbors=10)
ann_model = MLPClassifier(hidden_layer_sizes=(70, 40), activation="tanh", learning_rate_init=0.01, validation_fraction=0.1, n_iter_no_change=80, max_iter=10000) # Increase max_iter from 1000 to ensure convergence
svm_model = SVC(kernel="rbf", degree=3, C=1.0, gamma="scale")

# Create a list of tuples (name, model) for the base models
base_models = [
    ("DecisionTree", dt_model),
    ("kNN", knn_model),
    ("ANN", ann_model),
    ("SVM", svm_model)
]

# Choose a final estimator for the stacking ensemble
# Logistic Regression is a common choice for combining predictions
final_estimator = LogisticRegression()

# Create the stacking ensemble
ensemble = StackingClassifier(estimators=base_models, final_estimator=final_estimator)

# Train the ensemble model
fit!(ensemble, train_inputs, train_targets)

# Evaluate the ensemble model
model_accuracy = score(ensemble, test_inputs, test_targets)
println("Ensemble model accuracy: $(model_accuracy * 100) %")

# Save the model
#JLD.save("ensemble.jld", "model", ensemble)


Ensemble model accuracy: 60.430798735660765 %


In [17]:
"""
using Plots

# Define the data for each model
ann_means = [0.947, 0.947, 0.925, 0.788, 0.948, 0.933, 0.8, 0.9]
ann_stds = [0.018, 0.04, 0.07, 0.097, 0.036, 0.039, 0.4, 0.3]
svm_means = [0.947, 0.947, 0.927, 0.953, 0.94, 0.4, 0.5]
svm_stds = [0.03, 0.038, 0.092, 0.051, 0.043, 0.082, 0.7]
dt_means = [0.927, 0.913, 0.913, 0.913, 0.913, 0.913]
dt_stds = [0.043, 0.045, 0.045, 0.045, 0.045, 0.045]
knn_means = [0.947, 0.947, 0.96, 0.94, 0.913, 0.507]
knn_stds = [0.038, 0.038, 0.015, 0.043, 0.104, 0.068]

# Create subplots for each model
p1 = bar(1:6, ann_means, yerr=ann_stds, title="ANN", legend=false)
p2 = bar(1:6, svm_means, yerr=svm_stds, title="SVM", legend=false)
p3 = bar(1:6, dt_means, yerr=dt_stds, title="Decision Tree", legend=false)
p4 = bar(1:6, knn_means, yerr=knn_stds, title="KNN", legend=false)

# Customize the y-axis and labels
for p in [p1, p2, p3, p4]
    ylabel!(p, "Accuracy")
    xlabel!(p, "Set of Hyperparameters")
end

# Combine the plots into one figure
plot(p1, p2, p3, p4, layout=(2,2), size=(800,600))
"""

"using Plots\n\n# Define the data for each model\nann_means = [0.947, 0.947, 0.925, 0.788, 0.948, 0.933, 0.8, 0.9]\nann_stds = [0.018, 0.04, 0.07, 0.097, 0.036, 0.039, 0.4, 0.3]\nsvm_means = [0.947, 0.947, 0.927, 0.953, 0.94, 0.4, 0.5]\nsvm_stds = [0.03, 0.038, 0.092, 0.051, 0." ⋯ 492 bytes ⋯ "means, yerr=knn_stds, title=\"KNN\", legend=false)\n\n# Customize the y-axis and labels\nfor p in [p1, p2, p3, p4]\n    ylabel!(p, \"Accuracy\")\n    xlabel!(p, \"Set of Hyperparameters\")\nend\n\n# Combine the plots into one figure\nplot(p1, p2, p3, p4, layout=(2,2), size=(800,600))\n"

## Copy-pasted from unit 6 but something similar would go here ##

### Best model configuration

Based on the hyperparameters provided for each model and the results we have, the best configurations for each model according to **accuracy** are:


- **Artificial Neural Network (ANN)**: The best-performing ANN model uses the architecture [100, 100, 100] with 'relu' activation, a learning rate of 0.01, a validation ratio of 0.1, and a maximum of 1000 iterations. This suggests that a more complex model with a higher number of neurons was able to capture the complexity of the data better than simpler models.

- **Support Vector Machine (SVM)**: The SVM model that performed best had the 'linear' kernel with a C value of 1.0 and 'auto' gamma setting. This indicates a model that balances margin and misclassification error, without the need for the complexity of a non-linear kernel.

- **Decision Tree**: The best decision tree model had a maximum depth of 3. This suggests that a simpler model, which avoids overfitting by not going too deep into the tree, was sufficient to capture the relevant patterns in the data.

- **K-Nearest Neighbors (KNN)**: The KNN model that yielded the best results had 15 neighbors. This points to an intermediate complexity that balances between smoothing out the noise and capturing sufficient detail from the dataset.

We shall see their performance in the following table:


| Best Model Configuration          | Accuracy         | Sensitivity      | Specificity      | PPV              | NPV              | F_Score          | Err_Rate         |
|---------------------|------------------|------------------|------------------|------------------|------------------|------------------|------------------|
| ANN (Model 5)       | 0.948 ± 0.036    | 0.948 ± 0.036    | 0.979 ± 0.013    | 0.959 ± 0.027    | 0.969 ± 0.025    | 0.948 ± 0.036    | 0.052 ± 0.036    |
| SVM (Model 4)       | 0.953 ± 0.051    | 0.953 ± 0.051    | 0.980 ± 0.022    | 0.964 ± 0.034    | 0.973 ± 0.033    | 0.953 ± 0.051    | 0.047 ± 0.051    |
| Decision Tree (Model 1) | 0.927 ± 0.043 | 0.927 ± 0.043    | 0.967 ± 0.021    | 0.940 ± 0.032    | 0.955 ± 0.032    | 0.927 ± 0.043    | 0.073 ± 0.043    |
| KNN (Model 3)       | 0.96 ± 0.015     | 0.96 ± 0.015     | 0.98 ± 0.009     | 0.965 ± 0.011    | 0.978 ± 0.015    | 0.96 ± 0.015     | 0.04 ± 0.015     |

### Best performing model 

The KNN model outperforms the others with the highest accuracy of 0.96 and the lowest standard deviation (0.015). It appears to offer the best compromise between bias and variance, which is a key factor in its superior performance in this evaluation. It's worth noting, though, that the ideal model could vary depending on the specific data and the context of the problem. While accuracy is a critical metric, and we believe is a good metric for our use case, it may not be the sole criterion for success in other cases, especially when the consequences of false positives and false negatives differ significantly. In such scenarios, other metrics like sensitivity or specificity may be more relevant for optimization.

# ---------------------------------------------- #